### PR TITLE
Loading plugins with strapi-server.js

### PR DESCRIPTION
**Pull request by petersg83**
First draft of plugin load

### CODE REVIEW

File: packages/core/strapi/lib/core/plugins/create-plugin.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/plugins/create-plugin.js:17]: we could reverse the logic like so. wdyt ?

`strapi-server.js`
```js
module.exports = strapi => {

  return {
    config: strapi.loaders.loadConfig('./config'), 
  }
}
```
that would allow reusing the same logic as in the app-configuration loader and allow loading an entire config folder more easily. not sure that's necessary though

**Review by alexandrebodin — COMMENTED** [review]
This is goign in the right direction 👍

File: packages/core/strapi/lib/core/plugins/create-plugin.js
Comment (petersg83) [reply packages/core/strapi/lib/core/plugins/create-plugin.js:17]: I see, however if we are going this way we would not need to return `{ config, contentTypes }` at the end, right ?
Then it would only be a registration logic, not a declarative one. It's good for me but you were not really in favor of it if I remember correctly. wdyt ?

Comment (alexandrebodin) [reply packages/core/strapi/lib/core/plugins/create-plugin.js:17]: actually it would only help the developer load the config but we still need to return them.

File: packages/connectors/bookshelf/lib/models.js
Comment (alexandrebodin) [inline packages/connectors/bookshelf/lib/models.js:56]: the connector is getting remove I think you didn't want to rename it back to bookself :)

File: packages/core/strapi/lib/middlewares/router/utils/routerChecker.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/middlewares/router/utils/routerChecker.js:48]: Can you explain this change ?

File: packages/core/strapi/lib/core/registries/services.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/services.js:22]: maybe throwing would be better suited

File: packages/core/strapi/lib/core/registries/middlewares.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/middlewares.js:6]: _ isn't necessary here to stay consistent

File: packages/core/strapi/lib/core/registries/content-types.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/content-types.js:6]: Should we finish this before merging ?

File: packages/core/strapi/lib/core/registries/config.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/registries/config.js:9]:
```suggestion
    ..._config, // TODO: remove
```

File: packages/core/strapi/lib/core/loaders/load-modules.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/load-modules.js:35]: You can delete this right ?

File: packages/core/strapi/lib/core/loaders/load-middlewares.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/load-middlewares.js:7]: TODO after merging: We will need to use the middleware registry instead so we put everything into it and use it in the initializeMiddlewares later on

File: packages/core/strapi/lib/core/loaders/load-extensions.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/load-extensions.js:7]: are you still using this module ?

File: packages/core/content-manager/services/metrics.js
Comment (alexandrebodin) [inline packages/core/content-manager/services/metrics.js:30]: You can move the function in the factory to simplify

File: packages/core/content-type-builder/server/middlewares/index.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/server/middlewares/index.js:1]: We can remove this file

File: packages/core/content-type-builder/server/policies/index.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/server/policies/index.js:1]: We can remove this file

File: packages/core/content-type-builder/services/content-types.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/services/content-types.js:242]: do remove I guess

File: packages/core/content-type-builder/services/content-types.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/services/content-types.js:52]: you can delete

File: packages/core/content-type-builder/services/schema-builder/index.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/services/schema-builder/index.js:161]:
```suggestion
          schema => schema.flush()
```

File: packages/core/content-type-builder/services/schema-builder/index.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/services/schema-builder/index.js:156]: Any reason you added async ?

File: packages/core/database.old/lib/connector-registry.js
Comment (alexandrebodin) [inline packages/core/database.old/lib/connector-registry.js:52]: this code isn't used anymore lookis like you don't need to change anything in database.old

File: packages/core/email/server/middlewares/index.js
Comment (alexandrebodin) [inline packages/core/email/server/middlewares/index.js:1]: Looks like we can remove this file

File: packages/core/email/server/policies/index.js
Comment (alexandrebodin) [inline packages/core/email/server/policies/index.js:1]: Looks like we can remove this file

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:28]: you can remove

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:89]: Can you document what is left to do to be able to just have plugins here

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:244]: you can clean this up

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:106]: Does this really work ?

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:357]: can be removed it seems

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:363]: can be removed it seems

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:385]:
```suggestion
    await this.container.get('modules')[lifecycleName]()
```

File: packages/core/strapi/lib/commands/build.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/build.js:18]: shouldn't this be config.get ?

File: packages/core/strapi/lib/commands/develop.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/develop.js:23]: Same should the getOr be config.get ?

File: packages/core/strapi/lib/commands/watchAdmin.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/watchAdmin.js:18]: same for config.get instead of getOr ?

File: packages/core/strapi/lib/core/app-configuration/index.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/app-configuration/index.js:95]: feels like not keep createConfigProvider make the usage in the commands inconsistent. we might want to rethink this

File: packages/core/strapi/lib/core/app-configuration/index.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/app-configuration/index.js:14]: Do we really need to comment these ?

File: packages/core/strapi/lib/core/domain/module/index.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/domain/module/index.js:12]: We need to review this eventually so the logic for loaded is more robust here.

File: packages/core/strapi/lib/core/fs.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/fs.js:54]: you can use fse.appendFile instead to keep using fs-extra

File: packages/core/strapi/lib/core/loaders/load-modules.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/core/loaders/load-modules.js:25]: this can be removed ?

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Made a first global review

File: packages/plugins/users-permissions/config/functions/bootstrap.js
Comment (petersg83) [reply packages/plugins/users-permissions/config/functions/bootstrap.js:36]: Yes! Funny how prettier handled it

File: packages/plugins/users-permissions/admin/src/utils/tests/formatPolicies.test.js
Comment (petersg83) [reply packages/plugins/users-permissions/admin/src/utils/tests/formatPolicies.test.js:5]: To have a consistant format. It is all lowercased sometimes and camelCased some other times

File: packages/plugins/i18n/server/register.js
Comment (petersg83) [reply packages/plugins/i18n/server/register.js:15]: The integration with graphql will need to be reviewed, in this case the config of a plugin is not located there anymore, the line would not work

File: packages/plugins/i18n/server/register.js
Comment (petersg83) [reply packages/plugins/i18n/server/register.js:41]: It doesn't work anymore. I'm waiting for the graphql branch to be merged before fixing it

File: packages/plugins/i18n/server/middlewares/i18n/index.js
Comment (petersg83) [reply packages/plugins/i18n/server/middlewares/i18n/index.js:13]: It would work for now but not sure how we will handle routes in V4

File: packages/core/strapi/lib/middlewares/router/utils/routerChecker.js
Comment (petersg83) [reply packages/core/strapi/lib/middlewares/router/utils/routerChecker.js:48]: Seems a mistake

File: packages/core/strapi/lib/core/registries/services.js
Comment (petersg83) [reply packages/core/strapi/lib/core/registries/services.js:22]: In the other registres it is like that:
```
    get(policyUID) {
      return policies[policyUID];
    },
```

So returning undefined.
I can throw an error but would need to do it also for the others for consistency

Comment (alexandrebodin) [reply packages/core/strapi/lib/core/registries/services.js:22]: Yeah agreed we can do that a bit later :)

File: packages/core/strapi/lib/Strapi.js
Comment (petersg83) [reply packages/core/strapi/lib/Strapi.js]: Still needed?

File: packages/core/strapi/lib/Strapi.js
Comment (petersg83) [reply packages/core/strapi/lib/Strapi.js:108]: Worked for me. Here is the code of `stopWithError`;
```js
  stopWithError(err, customMessage) {
    this.log.debug(`⛔️ Server wasn't able to start properly.`);
    if (customMessage) {
      this.log.error(customMessage);
    }

    this.log.error(err);
    return this.stop();
  }
```

Why wouldn't it work?

File: packages/core/strapi/lib/commands/build.js
Comment (petersg83) [reply packages/core/strapi/lib/commands/build.js:18]: No because it comes from `loadConfiguration(dir)` and not `createConfigProvider`.
In Strapi.js we have:
```js
    const appConfig = loadConfiguration(this.dir, opts);
    this.container.register('config', createConfigProvider(appConfig));
```

File: packages/core/strapi/lib/commands/develop.js
Comment (petersg83) [reply packages/core/strapi/lib/commands/develop.js:23]: see answer above

File: packages/core/strapi/lib/commands/watchAdmin.js
Comment (petersg83) [reply packages/core/strapi/lib/commands/watchAdmin.js:18]: see answer above

File: packages/core/strapi/lib/core/app-configuration/index.js
Comment (petersg83) [reply packages/core/strapi/lib/core/app-configuration/index.js:95]: Yes we can rethink it, after merging?

File: packages/core/strapi/lib/core/domain/module/index.js
Comment (petersg83) [reply packages/core/strapi/lib/core/domain/module/index.js:12]: we can use an object instead

File: packages/core/strapi/lib/core/fs.js
Comment (petersg83) [reply packages/core/strapi/lib/core/fs.js:54]: I don't find it in their doc : https://github.com/jprichardson/node-fs-extra

File: packages/core/strapi/lib/core/registries/content-types.js
Comment (petersg83) [reply packages/core/strapi/lib/core/registries/content-types.js:6]: I put a todo. I would prefer to handle it a bit later if it's ok for you

File: packages/core/content-type-builder/server/services/schema-builder/index.js
Comment (alexandrebodin) [inline packages/core/content-type-builder/server/services/schema-builder/index.js:161]: even better

File: packages/core/strapi/lib/core/fs.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/core/fs.js:54]: that's just because it's not documented but they export all the same function as fs so they didn't bother document as they have the same API just promisified: https://github.com/jprichardson/node-fs-extra/blob/master/lib/fs/index.js#L9

File: packages/core/strapi/lib/core/domain/module/index.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/core/domain/module/index.js]: That's ok for now but the issue is we don't garantee that you can't call bootstrap if registered has been called for example :) that's would require a bit more state maangement to implement and it doesn't seem too critical though

File: packages/core/strapi/lib/Strapi.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/Strapi.js]: Nope :)

**Pull request merged**
